v2/tools/asoctl/internal/crd/cleaner.go

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package crd
+
+import (
+	"context"
+	"regexp"
+	"time"
+
+	"github.com/Azure/azure-service-operator/v2/internal/logging"
+	"github.com/pkg/errors"
+
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Cleaner struct {
+	apiExtensionsClient apiextensionsclient.CustomResourceDefinitionInterface
+	client              client.Client
+	migrationBackoff    wait.Backoff
+	dryRun              bool
+}
+
+func NewCleaner(apiExtensionsClient apiextensionsclient.CustomResourceDefinitionInterface, client client.Client, dryRun bool) *Cleaner {
+	migrationBackoff := wait.Backoff{ // TODO: Still need to see if we want exponential backoff or linear is fine
+		Duration: 2 * time.Second, // wait 2s between attempts, this will help us in a state of conflict.
+		Steps:    3,               // 3 retry on error attempts per object
+	}
+	return &Cleaner{
+		apiExtensionsClient: apiExtensionsClient,
+		client:              client,
+		migrationBackoff:    migrationBackoff,
+		dryRun:              dryRun,
+	}
+}
+
+func (c *Cleaner) Run(ctx context.Context) error {
+	list, err := c.apiExtensionsClient.List(ctx, v1.ListOptions{})
+	if err != nil {
+		return errors.Wrap(err, "failed to list CRDs")
+	}
+
+	if list == nil || len(list.Items) == 0 {
+		return errors.New("found 0 results, make sure you have ASO CRDs installed")
+
+	}
+
+	var updated int
+	crdRegexp := regexp.MustCompile(`.*\.azure\.com`)
+	deprecatedVersionRegexp := regexp.MustCompile(`v1alpha1api\d{8}(preview)?(storage)?`)
+	for _, crd := range list.Items {
+		crd := crd
+
+		if !crdRegexp.MatchString(crd.Name) {
+			continue
+		}
+
+		newStoredVersions, matchedStoredVersion := removeMatchingStoredVersions(crd.Status.StoredVersions, deprecatedVersionRegexp)
+
```

Comment:
I thought you were going to split these two cases:

`len(newStoredVersions) == 0` becomes an error because that shouldn't ever happen and if it does we should just abort the whole run right now given the user is probably trying to run this tool against a cluster w/ the wrong version of ASO installed.

`len(newStoredVersions) != len(crd.Status.StoredVersions)` means there was nothing to remove (already removed from this CRD), so your `klog.Infof("nothing to update")` is correct